src/native/libs/System.Native/pal_io.c

```diff
@@ -1952,17 +1952,34 @@ int32_t SystemNative_PWrite(intptr_t fd, void* buffer, int32_t bufferSize, int64
 #if (HAVE_PREADV || HAVE_PWRITEV) && !defined(TARGET_WASM)
 static int GetAllowedVectorCount(int32_t vectorCount)
 {
-    int allowedCount = (int)vectorCount;
+#if defined(_SC_IOV_MAX)
+    static volatile int s_iovMax = 0;
 
-#if defined(IOV_MAX)
-    if (IOV_MAX < allowedCount)
+    int iovMax = s_iovMax;
+    if (iovMax == 0)
     {
-        // We need to respect the limit of items that can be passed in iov.
-        // In case of writes, the managed code is reponsible for handling incomplete writes.
-        // In case of reads, we simply returns the number of bytes read and it's up to the users.
-        allowedCount = IOV_MAX;
+        // For macOS arm64 the IOV_MAX reports 1024, but fails with EINVAL for such inputs.
```

Comment:
@tmds @am11 @MichalPetryka @janvorli I am having hard time getting this to work on macOS arm64 and I need help to understand why. `IOV_MAX` reports 1024, but `pwritev` fails with `EINVAL` ([logs](https://helixr1107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-109826-merge-db79198bc31747eeab/System.IO.FileSystem.Tests/1/console.2a94267f.log?helixlogtype=result)) for 1024 vectors. When I lower the value to sth around 500, it works just fine ([logs](https://helixr1107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-109826-merge-a977e0f632e44b7d9d/System.IO.FileSystem.Tests/1/console.32a67139.log?helixlogtype=result)).

I don't have an access to arm64 macOS machine so I've asked @jozkee to run [this app](https://gist.github.com/adamsitnik/49361471bf6415ad0c78c6ddcc6f3313) for me and it worked fine:

```log
gcc -o repro repro.c
./repro             
IOV_MAX=1024 
ok writev=1024 
ok pwritev=1024
```

I could hardcode the value for macOS arm64, but I would prefer to understand what is going on here. Is it possible that with .NET Runtime I call the sys-call directly and with the `.c` program I call some libc wrapper that does something similar to what is described in the [man](https://linux.die.net/man/2/pwritev):

> the glibc wrapper functions do some extra work if they detect that the underlying kernel system call failed because this limit was exceeded. In the case of readv() the wrapper function allocates a temporary buffer large enough for all of the items specified by iov, passes that buffer in a call to [read](https://linux.die.net/man/2/read)(2), copies data from the buffer to the locations specified by the iov_base fields of the elements of iov, and then frees the buffer. The wrapper function for writev() performs the analogous task using a temporary buffer and a call to [write](https://linux.die.net/man/2/write)(2).

Reply:
From https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/writev.2.html:

>      [EINVAL]           The sum of the iov_len values in the iov array over-flows overflows a 32-bit integer.

Perhaps the test is hitting this limit.

Reply:
I think that @tmds is right and that's a plausible explanation.

Reply:
@tmds Thanks a lot!